File: 46.permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  const len = nums.length;
  const res = [];
  function travel(level) {
    if (level === len) {
      res.push(Array.from(nums));
      return;
    }
    for (let i = 0; i <= level; ++i) {
      swap(i, level);
      travel(level + 1);
      swap(i, level);
    }
  }

  travel(1);

  function swap(i, j) {
    const t = nums[i];
    nums[i] = nums[j];
    nums[j] = t;
  }

  // console.log("res", res);
  return res;
};

// permute([0, 1]);
permute([1, 2, 3]);

File: 0/46.permutations.py

from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        print(123)
        res = []
        def recursive(level):
            if level == len(nums):
                res.append(nums[::])
                return
            for i in range(0, level + 1):
                nums[i], nums[level] = nums[level], nums[i]
                recursive(level + 1)
                nums[level], nums[i] = nums[i], nums[level]
            return res

        recursive(nums)
        return res
